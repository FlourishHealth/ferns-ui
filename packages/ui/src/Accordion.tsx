import FontAwesome6 from "@expo/vector-icons/FontAwesome6";
import React, {FC, useEffect, useState} from "react";
import {Pressable, View} from "react-native";

import {AccordionProps} from "./Common";
import {Heading} from "./Heading";
import {InfoModalIcon} from "./InfoModalIcon";
import {Text} from "./Text";
import {useTheme} from "./Theme";

export const Accordion: FC<AccordionProps> = ({
  children,
  isCollapsed = false,
  title,
  subtitle,
  includeInfoModal = false,
  infoModalChildren,
  infoModalSubtitle,
  infoModalText,
  infoModalTitle,
}) => {
  const {theme} = useTheme();
  const [collapsed, setCollapsed] = useState(false);

  // The external collapse state should override the internal collapse state.
  useEffect(() => {
    setCollapsed(isCollapsed);
  }, [isCollapsed]);

  return (
    <View
      style={{
        padding: 16,
        borderBottomColor: theme.border.default,
        borderTopColor: theme.border.default,
        borderTopWidth: 2,
        borderBottomWidth: 2,
        width: "100%",
      }}
    >
      <View style={{flexDirection: "row", justifyContent: "space-between", alignItems: "center"}}>
        <View style={{flexDirection: "column", gap: 4}}>
          <View style={{flexDirection: "row", alignItems: "center"}}>
            <Heading>{title}</Heading>
            {includeInfoModal && infoModalTitle && (
              <InfoModalIcon
                infoModalChildren={infoModalChildren}
                infoModalSubtitle={infoModalSubtitle}
                infoModalText={infoModalText}
                infoModalTitle={infoModalTitle}
              />
            )}
          </View>
          {subtitle && <Text>{subtitle}</Text>}
        </View>
        <View>
          <Pressable
            aria-role="button"
            hitSlop={{top: 20, bottom: 20, left: 20, right: 20}}
            onPress={() => setCollapsed(!collapsed)}
          >
            <FontAwesome6
              color={theme.text.link}
              name={collapsed ? "chevron-down" : "chevron-up"}
              selectable={undefined}
              size={16}
            />
          </Pressable>
        </View>
      </View>
      {collapsed ? null : <View style={{marginTop: 8}}>{children}</View>}
    </View>
  );
};
