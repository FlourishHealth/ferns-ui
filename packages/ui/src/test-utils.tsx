import {render} from "@testing-library/react-native";
import React from "react";
import {ThemeProvider} from "./Theme";

export const renderWithTheme = (ui: React.ReactElement) => {
  return render(<ThemeProvider>{ui}</ThemeProvider>);
};

export const createCommonMocks = () => ({
  onChange: jest.fn(),
  onFocus: jest.fn(),
  onBlur: jest.fn(),
  onEnter: jest.fn(),
  onSubmitEditing: jest.fn(),
  onIconClick: jest.fn(),
});

export const setupComponentTest = () => {
  jest.useFakeTimers();
  return createCommonMocks();
};

export const teardownComponentTest = () => {
  jest.useRealTimers();
  jest.clearAllMocks();
};
