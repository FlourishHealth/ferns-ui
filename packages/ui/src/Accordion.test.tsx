import {fireEvent} from "@testing-library/react-native";
import React from "react";
import {Text} from "react-native";

import {Accordion} from "./Accordion";
import {renderWithTheme} from "./test-utils";

describe("Accordion", () => {
  const defaultProps = {
    title: "Test Title",
    children: <Text testID="test-content">Test Content</Text>,
  };

  it("renders correctly with default props", () => {
    const {toJSON} = renderWithTheme(<Accordion {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders title and children", () => {
    const {getByText, getByTestId} = renderWithTheme(
      <Accordion {...defaultProps} isCollapsed={false} />
    );
    expect(getByText("Test Title")).toBeTruthy();
    expect(getByTestId("test-content")).toBeTruthy();
  });

  it("renders subtitle when provided", () => {
    const {getByText} = renderWithTheme(<Accordion {...defaultProps} subtitle="Test Subtitle" />);
    expect(getByText("Test Subtitle")).toBeTruthy();
  });

  it("collapses when isCollapsed is true", () => {
    const {queryByTestId} = renderWithTheme(<Accordion {...defaultProps} isCollapsed />);
    expect(queryByTestId("test-content")).toBeNull();
  });

  it("expands when isCollapsed is false", () => {
    const {getByTestId} = renderWithTheme(<Accordion {...defaultProps} isCollapsed={false} />);
    expect(getByTestId("test-content")).toBeTruthy();
  });

  it("toggles visibility when header is pressed", () => {
    const {queryByTestId, getByTestId} = renderWithTheme(
      <Accordion {...defaultProps} isCollapsed={false} />
    );

    // Content should be visible initially
    expect(getByTestId("test-content")).toBeTruthy();

    // Click the toggle button
    const toggleButton = getByTestId("accordion-toggle");
    fireEvent.press(toggleButton);

    // Content should be hidden after click
    expect(queryByTestId("test-content")).toBeNull();
  });

  it("renders info modal icon when includeInfoModal is true and infoModalTitle is provided", () => {
    const {getByTestId} = renderWithTheme(
      <Accordion {...defaultProps} includeInfoModal infoModalTitle="Test Info Title" />
    );

    // The info modal icon should be present
    expect(getByTestId("info-icon")).toBeTruthy();
  });

  it("toggles content visibility when header is clicked", () => {
    const {getByTestId, queryByText, getByText} = renderWithTheme(
      <Accordion {...defaultProps} isCollapsed={false} />
    );

    // Content should be visible initially
    expect(queryByText("Test Content")).toBeTruthy();

    // Click the header to collapse
    const toggleButton = getByTestId("accordion-toggle");
    fireEvent.press(toggleButton);

    // Content should be hidden after click
    expect(queryByText("Test Content")).toBeNull();

    // Click again to expand
    fireEvent.press(toggleButton);
    expect(getByText("Test Content")).toBeTruthy();
  });

  it("renders info modal icon when includeInfoModal is true and infoModalTitle is provided", () => {
    const {getByTestId} = renderWithTheme(
      <Accordion {...defaultProps} includeInfoModal infoModalTitle="Info Title" />
    );

    // The info modal icon should be present
    expect(getByTestId("info-icon")).toBeTruthy();
  });

  it("does not render info modal icon when includeInfoModal is false", () => {
    const {queryByTestId} = renderWithTheme(
      <Accordion {...defaultProps} includeInfoModal={false} infoModalTitle="Info Title" />
    );

    // The info modal icon should not be present
    expect(queryByTestId("info-icon")).toBeNull();
  });
});
