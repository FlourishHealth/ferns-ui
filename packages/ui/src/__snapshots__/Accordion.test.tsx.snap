// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders correctly with default props 1`] = `
<View
  style={
    {
      "borderBottomColor": "#CDCDCD",
      "borderBottomWidth": 2,
      "borderTopColor": "#CDCDCD",
      "borderTopWidth": 2,
      "padding": 16,
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "gap": 4,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          numberOfLines={0}
          style={
            {
              "color": "#1C1C1C",
              "fontFamily": "heading-bold",
              "fontSize": 16,
            }
          }
        >
          Test Title
        </Text>
      </View>
    </View>
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        aria-role="button"
        collapsable={false}
        focusable={true}
        hitSlop={
          {
            "bottom": 20,
            "left": 20,
            "right": 20,
            "top": 20,
          }
        }
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="accordion-toggle"
      >
        <View
          color="#0A7092"
          name="chevron-up"
          size={16}
          testID="chevron-up"
        />
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginTop": 8,
      }
    }
  >
    <Text
      testID="test-content"
    >
      Test Content
    </Text>
  </View>
</View>
`;
